bypros/api.go

```diff
@@ -0,0 +1,213 @@
+package bypros
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gorilla/websocket"
+	"go.dedis.ch/cothority/v3"
+	"go.dedis.ch/cothority/v3/skipchain"
+	"go.dedis.ch/onet/v3"
+	"go.dedis.ch/onet/v3/network"
+	"go.dedis.ch/onet/v4/log"
+	"go.dedis.ch/protobuf"
+	"golang.org/x/xerrors"
+)
+
+var overlayClient OverlayClient = onetOverlay{
+	onet.NewClient(cothority.Suite, ServiceName),
+}
+
+// OverlayClient defines the primitives needed by our client. That's all we need
+// from the outer world.
+type OverlayClient interface {
+	SendProtobuf(dst *network.ServerIdentity, msg interface{}, ret interface{}) error
+	OpenWS(url string) (WsHandler, error)
+}
+
+// WsHandler defines the primitives to handle a websocket connection.
+type WsHandler interface {
+	Close() error
+	Write(messageType int, data []byte) error
+	Read() (messageType int, p []byte, err error)
+}
+
+// NewClient creates a new proxy client
+func NewClient() *Client {
+	return &Client{OverlayClient: overlayClient}
+}
+
+// Client defines a proxy client
+//
+// - implements OverlayClient
+type Client struct {
+	OverlayClient
+}
+
+// Follow sends a request to start following a node, ie. listening to new blocks
+// and updating the database accordingly.
+func (c *Client) Follow(host, target *network.ServerIdentity,
+	scID skipchain.SkipBlockID) error {
+
+	req := Follow{
+		Target: target,
+		ScID:   scID,
+	}
+
+	resp := EmptyReply{}
+
+	err := c.SendProtobuf(host, &req, &resp)
+	if err != nil {
+		return xerrors.Errorf("failed to send follow request: %v", err)
+	}
+
+	return nil
+}
+
+// UnFollow sends a request to stop following a node.
+func (c *Client) UnFollow(host *network.ServerIdentity) error {
+	req := UnFollow{}
+	resp := EmptyReply{}
+
+	err := c.SendProtobuf(host, &req, &resp)
+	if err != nil {
+		return xerrors.Errorf("failed to send follow request: %v", err)
+	}
+
+	return nil
+}
+
+// Query sends a query request. The query must be a read-only SQL query, for
+// example:
+//   Select * from cothority.block
+func (c *Client) Query(host *network.ServerIdentity, query string) ([]byte, error) {
+	req := Query{
+		Query: query,
+	}
+	resp := QueryReply{}
+
+	err := c.SendProtobuf(host, &req, &resp)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to send query request: %v", err)
+	}
+
+	return resp.Result, nil
+}
+
+// CatchUP sends a request to catch up from a particular block. A catch up will
+// update the database from the specified block until the end of the chain.
+// Every specifies the interval at which updates are sent.
+func (c *Client) CatchUP(ctx context.Context, host, target *network.ServerIdentity,
```

Comment:
I thought this would happen automagically. Once the proxy starts, it connects to the `StreamTransactions` byzcoin endpoint and then catches up.

Reply:
That was initially the case. It came out that splitting it into a separate service call makes the service much more manageable, from the developer and client's side. You can start following a node, then if you know that you *might* not be in sync, you can run a catchup in the background. You can also send a catchup request at some point if you just want to be sure that you are in sync.

Reply:
As I wrote above, following one node is not what you should do. To follow the chain, you need to follow the roster. The cothority-js does this by sending out a request to all nodes, and then keeps only the two fastest nodes for further request. If one of these nodes fails, it is put at the end of the list and the next node in the list is put in its place.

This allows for nodes to fail, and the system still to be manageable. Also, nodes might disappear altogether. If the client has to start catchup and give the node name, you put quite a big burden on him.

Reply:
not really feasible, that would require to know the node we want to follow at compile time, ie. when the service is included.

Reply:
Why do you need this on compile-time? here is what the ByzCoin service does on startup:
- load the config from the db
- tries to catchup from other nodes in the background
- starts operations

So when the bypros service receives a call to `Follow`, it can store the ID and the ServerIdentity in the DB offered from onet using `Context.Save`. When the bypros service is started the next time, it can use `Context.Load` to see if something is in the DB. If there is an ID and a ServerIdentity stored, it can call the catchup mechanism.